<!---High level numbered list of main steps of the workflow and hyperlink to any tools used. If multiple workflows/different modes perhaps have subheadings and numbered steps. Use nested numbering or bullets where required.--->
The workflow is composed of 2 subworkflows, each enabled by a command line option:

* [Small variant calling](#3-small-variant-calling-and-joint-genotyping): `--snp`
* [Structural variant calling](#5-structural-variant-calling-and-merging): `--sv`

### 1. Input and data preparation

The workflow requires:
1. A reference genome in [FASTA format](https://www.ncbi.nlm.nih.gov/genbank/fastaformat/)
2. Sequencing data for each of the trio samples in the form of a single [BAM file](https://samtools.github.io/hts-specs/SAMv1.pdf) or a folder of BAM files, either aligned or unaligned. If unaligned input is provided, alignment will be performed using [Minimap2](https://github.com/lh3/minimap2).
3. A [pedigree file](https://gatk.broadinstitute.org/hc/en-us/articles/360035531972-PED-Pedigree-format) to describe familial relationships between input samples. Sample names provided to the workflow should match exactly with their entries in the pedigree file. These names must be specified using the sample_name options: `proband_sample_name`, `pat_sample_name`, `mat_sample_name`. Note: the BAM header will be modified by the workflow so the RG ID matches the sample names provided with these options.

The workflow currently only supports the hg38/GRCh38 reference genome. When analysing human data, we recommend using [GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz](https://ont-exd-int-s3-euwst1-epi2me-labs.s3.amazonaws.com/ref/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz). For more information see [this blog post](https://lh3.github.io/2017/11/13/which-human-reference-genome-to-use) which outlines potential pitfalls with the various flavours of human references.

The input BAM file can be generated using our [wf-basecalling](https://github.com/epi2me-labs/wf-basecalling/) workflow, which is up to date with the current Dorado releases and models.

### 2. Data QC and pre-processing

The workflow starts by performing multiple checks of the input BAM file, as well as computing:
1. depth of sequencing with [mosdepth](https://github.com/brentp/mosdepth);
2. read alignment statistics with [fastcat](https://github.com/epi2me-labs/fastcat).

Results of these are available in the final report.

### 3. Small variant calling and joint genotyping.

The workflow implements a modified version of [Clair3-Nova](https://github.com/HKU-BAL/Clair3-Nova) to call germline variants in trio.
Clair3-Nova has been modified to improve parallelism in high-performance compute environments.
The workflow will select an appropriate Clair3 and Clair3-Nova model by detecting the basecall model from the input data.
If the input data does not have the required information to determine the basecall model, the workflow will require the basecall model to be provided explicitly with the `--override_basecaller_cfg` option. This will be used by the workflow to automatically determine the Clair3 and Clair3-Nova models to use for small variant calling.

The gVCFs generated by Clair3-Nova are then input in to [GLnexus](https://github.com/dnanexus-rnd/GLnexus) for joint genotyping.

Records in the VCF that are within a homopolymer region greater than 9 base pairs (`HP9`) will be annotated in both the individual and joint small variant VCF using a default BED file internal to the workflow.

### 4. Pedigree phasing 

The workflow can perform pedigree phasing of the variants by using the `--phased` option. The workflow uses [whatshap](https://whatshap.readthedocs.io/en/latest/guide.html#phasing-pedigrees) to perform pedigree phasing of the GLnexus joint variants VCF. Haplotype alleles of a proband are given as paternal|maternal; i.e. the first allele is the one inherited from the father and the second one the allele inherited from the mother. The BAMs will then be haplotagged using the phased VCFs.

### 5. Structural variant calling and merging

The workflow allows for merging of SVs using long-read sequencing data with [Sniffles2](https://github.com/fritzsedlazeck/Sniffles).
The workflow will perform SV calling on each individual and the results of these will be used in structural variant merging.
The SV workflow uses an appropriate tandem repeat annotation BED file to improve calling in repetitive regions. You can override the BED file used with the `--tr_bed` parameter. See the [Sniffles](https://github.com/fritzsedlazeck/Sniffles) documentation for more information.

Structural variants can be phased using `--phased`. However, this will cause the workflow to run small variant calling analysis, as SV phasing relies on the haplotagged reads generated in this stage.

Records in the VCF that are within a Tandem repeat (`TR`) region will be annotated in both the individual and joint structural variant VCF using a default BED file internal to the workflow.

### 6. Mendelian inheritance

[RTG Mendelian](https://realtimegenomics.github.io/rtg-tools/rtg_command_reference.html#mendelian) is used to report variants which do not follow Mendelian inheritance. This is run for small variant calling and structural variant multi-sample VCF.
